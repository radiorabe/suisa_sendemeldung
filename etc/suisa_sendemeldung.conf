# This is the configuration file for suisa_sendemeldung.py.
# - Values that are not commented out are examples and need to be replaced with
#   a valid value.
# - Values that are commented out are not required but can be overriden by this
#   config file.
# - All values can also be set by environment variables or with command line
#   arguments

# Access key for ACRCloud
access_key=abcdefghijklmnopqrstuvwxyzabcdef
# Stream Id in ACRCloud
stream_id=a-bcdefgh

# Specify whether to create a csv file
#csv=True

# Specify whether to send an email
#email=True
# Email sender
#email_from=sender@example.org
# Email recipient
#email_to=receiver@example.org
# SMTP server for sending the email
#email_server=smtp.example.org
# Password for sender on email server
#email_pass=HlZ5gPpHeugNpMk6
# Subject for the email
#email_subject=SUISA reporting
# Text for the email
#email_text=Attached the SUISA reporting

# Start date to fetch data from ACRCloud
#start_date=2018-10-01
# End date to fetch data from ACRCloud
#end_date=2018-10-31
# Specify whether to fetch the whole last month in data
#last_month=True
# Set timezone to use for conversion
#timezone = 'Europe/Zurich'

# The filename for writing to file and sending as email attachment
#filename=suisa_sendemeldung.csv

# Specify whether to print to stdout
#stdout=True
