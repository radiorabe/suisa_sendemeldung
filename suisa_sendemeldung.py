#!/usr/bin/env python3
from argparse import ArgumentParser
from csv import writer
from datetime import datetime, date, timedelta, timezone

import requests


class ACRClient:

    def __init__(self, access_key):
        self.access_key = access_key
        self.default_date = date.today()-timedelta(days=1)
        self.url = ('https://api.acrcloud.com/v1/'
                    'monitor-streams/{stream_id}/results')

    def get_data(self, stream_id, requested_date=None):
        if requested_date is None:
            requested_date = self.default_date
        url_params = dict(
            access_key=self.access_key,
            date=requested_date.strftime("%Y%m%d")
        )
        url = self.url.format(stream_id=stream_id)
        response = requests.get(url=url, params=url_params)
        response.raise_for_status()
        return response.json()

    def get_interval_data(self, stream_id, start, end):
        data = []
        while start <= end:
            data = data + client.get_data(args.stream_id, requested_date=start)
            start = start + timedelta(days=1)
        return data


def write_csv(filename, data):
    header = [
        'Sendedatum',
        'Sendezeit',
        'Sendedauer',
        'Titel',
        'KÃ¼nstler',
        'ISRC',
        'Label'
    ]
    with open(filename, mode='w') as csvfile:
        csvfile.write('sep=,\n')

        csv_writer = writer(csvfile, dialect='excel')
        csv_writer.writerow(header)

        for entry in data:
            metadata = entry.get('metadata')
            # parse timestamp (is in format "yyyy-mm-dd HH:MM:SS")
            timestamp = datetime.strptime(metadata.get('timestamp_utc'),
                                          '%Y-%m-%d %H:%M:%S')
            timestamp = (timestamp
                         .replace(tzinfo=timezone.utc)
                         .astimezone(tz=None))

            date = timestamp.strftime('%d/%m/%y')
            time = timestamp.strftime('%H:%M:%S')
            duration = timedelta(seconds=metadata.get('played_duration'))

            music = metadata.get('music')[0]
            title = music.get('title')
            artist = ', '.join([a.get('name') for a in music.get('artists')])
            if len(music.get('external_ids')) > 0:
                isrc = music.get('external_ids').get('isrc')
            else:
                isrc = ""
            label = music.get('label')

            csv_writer.writerow([date, time, duration,
                                 title, artist, isrc, label])


if __name__ == '__main__':
    parser = ArgumentParser(
                description='ACRCloud client for SUISA reporting @ RaBe.')
    parser.add_argument('--access_key',
                        help='the access key for ACRCloud (required)',
                        required=True)
    parser.add_argument('--stream_id',
                        help='the id of the stream at ACRCloud (required)',
                        required=True)
    parser.add_argument('--start_date',
                        help='the start date of the interval \
                              in format YYYY-MM-DD')
    parser.add_argument('--end_date',
                        help='the end date of the interval \
                              in format YYYY-MM-DD')
    parser.add_argument('--output',
                        help='file to write to (defaults to \
                              <script_name>_<date>.csv)')
    args = parser.parse_args()

    def print_error(msg):
        print('ERROR: {}\n'.format(msg))
        parser.print_help()
        quit(1)
    if not len(args.access_key) == 32:
        print_error('wrong format on access_key')
    if not args.end_date:
        end_date = date.today()
    else:
        end_date = datetime.strptime(args.end_date, '%Y-%m-%d').date()
    if not args.start_date:
        start_date = end_date - timedelta(days=30)
    else:
        start_date = datetime.strptime(args.start_date, '%Y-%m-%d').date()

    client = ACRClient(args.access_key)
    data = client.get_interval_data(args.stream_id, start_date, end_date)

    write_csv(args.output, data)
