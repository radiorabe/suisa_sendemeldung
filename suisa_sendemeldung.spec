#
# spec file for package suisa_sendemeldung
#
# Copyright (c) 2018 Radio Bern RaBe
#                    http://www.rabe.ch
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#
# Please submit enhancements, bugfixes or comments via GitHub:
# https://github.com/radiorabe/suisa_sendemeldung
#
%global srcname suisa_sendemeldung

%{?el7:%global python3_pkgversion 36}

Name:           %{srcname}
Version:        master
Release:        0%{?dist}
Summary:        ACRCloud client for SUISA reporting

License:        MIT
URL:            https://github.com/radiorabe/suisa_sendemeldung
Source0:        https://github.com/radiorabe/suisa_sendemeldung/archive/%{version}/%{name}-%{version}.tar.gz

BuildArch:      noarch
BuildRequires:  python%{python3_pkgversion}-configargparse
BuildRequires:  python%{python3_pkgversion}-devel
BuildRequires:  python%{python3_pkgversion}-setuptools
BuildRequires:  python%{python3_pkgversion}-requests
BuildRequires:  python3-devel
%{?systemd_requires}
BuildRequires:  systemd
Requires(pre):  shadow-utils
Requires:       python%{python3_pkgversion}-configargparse
Requires:       python%{python3_pkgversion}-requests
%{?python_enable_dependency_generator}

%description
ACRCloud client that fetches data on our playout history and
formats them in a CSV file format containing the data (like
Track, Title and, ISRC) requested by SUISA.
Also takes care of sending the report to SUISA via email for
hands-off operations.


%prep
%autosetup -n %{srcname}-%{version}

%build
%py3_build

%install
%py3_install
install -d %{buildroot}%{_unitdir}
install etc/systemd/%{srcname}.* %{buildroot}%{_unitdir}
install -d %{buildroot}%{_sysconfdir}
install etc/%{srcname}.conf %{buildroot}%{_sysconfdir}
install -d %{buildroot}%{_sysconfdir}/sysconfig
install etc/sysconfig/%{srcname} %{buildroot}%{_sysconfdir}/sysconfig

%check
%{__python3} setup.py test

%pre
getent group %{srcname} >/dev/null || groupadd -r %{srcname}
getent passwd %{srcname} >/dev/null || \
    useradd -r -g %{srcname} -d %{python3_sitelib}/%{srcname}/ -s /sbin/nologin \
    -c "SUISA ACRClient account" %{srcname}
exit 0

%post
%systemd_post %{srcname}.service
%systemd_post %{srcname}.timer

%preun
%systemd_preun %{srcname}.timer
%systemd_preun %{srcname}.service

%postun
%systemd_postun %{srcname}.timer
%systemd_postun %{srcname}.service

%files
%license LICENSE
%doc README.md
%{python3_sitelib}/%{srcname}/
%{python3_sitelib}/%{srcname}-*.egg-info/
%{_bindir}/%{srcname}
%{_unitdir}/%{srcname}.*
%config(noreplace)%{_sysconfdir}/%{srcname}.conf
%config(noreplace)%{_sysconfdir}/sysconfig/%{srcname}
