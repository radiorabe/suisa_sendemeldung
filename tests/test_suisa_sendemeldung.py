"""Test the suisa_sendemeldung.suisa_sendemeldung module."""
from datetime import date, datetime, timezone
from email.message import Message
from io import BytesIO
from unittest.mock import call, patch

from configargparse import ArgumentParser
from freezegun import freeze_time
from openpyxl import load_workbook
from pytest import mark

from suisa_sendemeldung import suisa_sendemeldung


def test_validate_arguments():
    """Test validate_arguments."""

    args = ArgumentParser()
    # length of bearer_token should be 32 or more chars long
    args.bearer_token = "iamclearlynotthirtytwocharslong"
    # check length of stream_id
    args.stream_id = "iamnot9chars"
    # one output option has to be set (but none is)
    args.file = False
    args.email = False
    args.stdout = False
    # last_month is in conflict with start_date and end_date
    args.last_month = True
    args.start_date = date(1993, 3, 1)
    with patch("suisa_sendemeldung.suisa_sendemeldung.ArgumentParser") as mock:
        suisa_sendemeldung.validate_arguments(mock, args)
        mock.error.assert_called_once_with(
            "\n"
            "- wrong format on bearer_token, expected larger than 32 characters but got 31\n"
            "- wrong format on stream_id, expected 9 characters but got 12\n"
            "- no output option has been set, specify one of --file, --email or --stdout\n"
            "- argument --last_month not allowed with --start_date or --end_date"
        )

    args.stdout = True
    args.filetype = "xlsx"
    with patch("suisa_sendemeldung.suisa_sendemeldung.ArgumentParser") as mock:
        suisa_sendemeldung.validate_arguments(mock, args)
        mock.error.assert_called_once_with(
            "\n"
            "- wrong format on bearer_token, expected larger than 32 characters but got 31\n"
            "- wrong format on stream_id, expected 9 characters but got 12\n"
            "- xlsx cannot be printed to stdout, please set --filetype to csv\n"
            "- argument --last_month not allowed with --start_date or --end_date"
        )


def test_parse_date():
    """Test parse_date."""

    # default, "last_month" case
    args = ArgumentParser()
    args.last_month = True
    with freeze_time("1996-04-01"):
        (start_date, end_date) = suisa_sendemeldung.parse_date(args)
    assert start_date == datetime(1996, 3, 1)
    assert end_date == datetime(1996, 3, 31)

    # if no start_date was set, default to 30 days before end_date
    args = ArgumentParser()
    args.last_month = False
    args.start_date = None
    args.end_date = "1996-03-31"
    (start_date, end_date) = suisa_sendemeldung.parse_date(args)
    assert start_date == date(1996, 3, 1)
    assert end_date == date(1996, 3, 31)

    # if no end_date was set, default to today
    args = ArgumentParser()
    args.last_month = False
    args.start_date = False
    args.end_date = False
    with freeze_time("1996-03-31"):
        (start_date, end_date) = suisa_sendemeldung.parse_date(args)
    assert start_date == date(1996, 3, 1)
    assert end_date == date(1996, 3, 31)

    # only specify start_date, selects up to today
    args = ArgumentParser()
    args.last_month = False
    args.start_date = "1996-03-01"
    args.end_date = False
    with freeze_time("1996-04-04"):
        (start_date, end_date) = suisa_sendemeldung.parse_date(args)
    assert start_date == date(1996, 3, 1)
    assert end_date == date(1996, 4, 4)


def test_parse_filename():
    """Test parse_filename."""

    # pass filename from cli
    args = ArgumentParser()
    args.filename = "/foo/bar"
    args.station_name_short = "test"
    filename = suisa_sendemeldung.parse_filename(args, None)
    assert filename == "/foo/bar"

    # last_month mode
    args = ArgumentParser()
    args.filename = None
    args.station_name_short = "test"
    args.last_month = True
    args.filetype = "xlsx"
    with freeze_time("1996-03-01"):
        filename = suisa_sendemeldung.parse_filename(args, datetime.now())
    assert filename == "test_1996_03.xlsx"

    # start date mode
    args = ArgumentParser()
    args.filename = None
    args.last_month = False
    args.start_date = "1996-03-01"
    args.filetype = "xlsx"
    args.station_name_short = "test"
    with freeze_time("1996-03-01"):
        filename = suisa_sendemeldung.parse_filename(args, datetime.now())
    assert filename == "test_1996-03-01.xlsx"


def test_check_duplicate():
    """Test check_duplicates."""

    # both records are music and not duplicate
    entry_a = {"metadata": {"music": [{"acrid": "123456789"}]}}
    entry_b = {"metadata": {"music": [{"acrid": "987654321"}]}}
    assert not suisa_sendemeldung.check_duplicate(entry_a, entry_b)

    # both records are music and duplicate
    entry_a = {"metadata": {"music": [{"acrid": "123456789"}]}}
    entry_b = {"metadata": {"music": [{"acrid": "123456789"}]}}
    assert suisa_sendemeldung.check_duplicate(entry_a, entry_b)

    # first record is custom and not duplicate
    entry_a = {"metadata": {"custom_files": [{"acrid": "123456789"}]}}
    entry_b = {"metadata": {"music": [{"acrid": "987654321"}]}}
    assert not suisa_sendemeldung.check_duplicate(entry_a, entry_b)

    # first record is custom and duplicate
    entry_a = {"metadata": {"custom_files": [{"acrid": "123456789"}]}}
    entry_b = {"metadata": {"music": [{"acrid": "123456789"}]}}
    assert suisa_sendemeldung.check_duplicate(entry_a, entry_b)

    # second record is custom and not duplicate
    entry_a = {"metadata": {"music": [{"acrid": "123456789"}]}}
    entry_b = {"metadata": {"custom_files": [{"acrid": "987654321"}]}}
    assert not suisa_sendemeldung.check_duplicate(entry_a, entry_b)

    # second record is custom and duplicate
    entry_a = {"metadata": {"music": [{"acrid": "123456789"}]}}
    entry_b = {"metadata": {"custom_files": [{"acrid": "123456789"}]}}
    assert suisa_sendemeldung.check_duplicate(entry_a, entry_b)

    # both records are custom and not duplicate
    entry_a = {"metadata": {"custom_files": [{"acrid": "123456789"}]}}
    entry_b = {"metadata": {"custom_files": [{"acrid": "987654321"}]}}
    assert not suisa_sendemeldung.check_duplicate(entry_a, entry_b)

    # both records are custom and duplicate
    entry_a = {"metadata": {"custom_files": [{"acrid": "123456789"}]}}
    entry_b = {"metadata": {"custom_files": [{"acrid": "123456789"}]}}
    assert suisa_sendemeldung.check_duplicate(entry_a, entry_b)


def test_merge_duplicates():
    """Test merge_duplicates."""

    # check if record_1 reduced to one record and durations are added together
    record_1 = {"metadata": {"music": [{"acrid": "123456789"}], "played_duration": 10}}
    record_2 = {"metadata": {"music": [{"acrid": "987654321"}], "played_duration": 10}}
    raw_records = [record_1, record_1, record_2]
    results = suisa_sendemeldung.merge_duplicates(raw_records)
    assert len(results) == 2
    assert results[0]["metadata"]["played_duration"] == 20


@mark.parametrize(
    "test_date,expected",
    [
        ("0000-00-00", ""),
    ],
)
def test_funge_release_date(test_date, expected):
    """Test funge_release_date."""

    results = suisa_sendemeldung.funge_release_date(test_date)
    assert results == expected


@patch("cridlib.get")
def test_get_csv(mock_cridlib_get):
    """Test get_csv."""
    mock_cridlib_get.return_value = "crid://rabe.ch/v1/test"

    # empty data
    data = []
    csv = suisa_sendemeldung.get_csv(data)
    # pylint: disable=line-too-long
    assert csv == (
        "Titel,Komponist,Interpret,Interpreten-Info,Sender,Sendedatum,Sendedauer,Sendezeit,Werkverzeichnisangaben,ISRC,Label,CD ID / Katalog-Nummer,Aufnahmedatum,Aufnahmeland,Erstveröffentlichungsdatum,Titel des Tonträgers (Albumtitel),Autor Text,Track Nummer,Genre,Programm,Bestellnummer,Marke,Label Code,EAN/GTIN,Identifikationsnummer\r\n"
    )
    # pylint: enable=line-too-long
    mock_cridlib_get.assert_not_called()

    # bunch of data
    mock_cridlib_get.reset_mock()
    data = [
        {
            "metadata": {
                "timestamp_local": "1993-03-01 13:12:00",
                "timestamp_utc": "1993-03-01 13:12:00",
                "played_duration": 60,
                "music": [{"title": "Uhrenvergleich", "acrid": "a1"}],
            }
        },
        {
            "metadata": {
                "timestamp_local": "1993-03-01 13:37:00",
                "timestamp_utc": "1993-03-01 13:37:00",
                "played_duration": 60,
                "custom_files": [
                    {
                        "acrid": "a2",
                        "title": "Meme Dub",
                        "artist": "Da Gang",
                        "album": "album, but string",
                        "contributors": {
                            "composers": [
                                "Da Composah",
                            ]
                        },
                        "release_date": "2023",
                        "external_ids": {"isrc": "DEZ650710376"},
                    }
                ],
            }
        },
        {
            "metadata": {
                "timestamp_local": "1993-03-01 16:20:00",
                "timestamp_utc": "1993-03-01 16:20:00",
                "played_duration": 60,
                "music": [
                    {
                        "acrid": "a3",
                        "title": "Bubbles",
                        "album": {
                            "name": "Da Alboom",
                        },
                        "release_date": "2022-12-13",
                        "artists": [
                            {
                                "name": "Mary's Surprise Act",
                            },
                            {
                                "name": "Climmy Jiff",
                            },
                        ],
                        "isrc": "DEZ650710376",
                        "label": "Jane Records",
                        "external_ids": {
                            "upc": "greedy-capitalist-number",
                        },
                    }
                ],
            }
        },
        {
            "metadata": {
                "timestamp_local": "1993-03-01 17:17:17",
                "timestamp_utc": "1993-03-01 17:17:17",
                "played_duration": 60,
                "custom_files": [
                    {
                        "acrid": "a4",
                        "artists": "Artists as string not list",
                    }
                ],
            }
        },
        {
            "metadata": {
                "timestamp_local": "1993-03-01 18:18:18",
                "timestamp_utc": "1993-03-01 18:18:18",
                "played_duration": 71337,
                "music": [{"title": "Long Playing", "acrid": "a5"}],
            }
        },
        {
            "metadata": {
                "timestamp_local": "1993-03-01 18:18:18",
                "timestamp_utc": "1993-03-01 18:18:18",
                "played_duration": 71337,
                "music": [
                    {
                        "title": "composer in works",
                        "acrid": "a6",
                        "works": [{"creators": [{"name": "Worker", "role": "W"}]}],
                    }
                ],
            }
        },
        {
            "metadata": {
                "timestamp_local": "1993-03-01 18:18:18",
                "timestamp_utc": "1993-03-01 18:18:18",
                "played_duration": 71337,
                "music": [
                    {
                        "title": "composer better in works",
                        "artists": [{"name": "same"}],
                        "contributors": {
                            "composers": ["same"],
                        },
                        "acrid": "a6",
                        "works": [{"creators": [{"name": "composer", "role": "C"}]}],
                    }
                ],
            }
        },
    ]
    csv = suisa_sendemeldung.get_csv(data, station_name="Station Name")
    # pylint: disable=line-too-long
    assert csv == (
        "Titel,Komponist,Interpret,Interpreten-Info,Sender,Sendedatum,Sendedauer,Sendezeit,Werkverzeichnisangaben,ISRC,Label,CD ID / Katalog-Nummer,Aufnahmedatum,Aufnahmeland,Erstveröffentlichungsdatum,Titel des Tonträgers (Albumtitel),Autor Text,Track Nummer,Genre,Programm,Bestellnummer,Marke,Label Code,EAN/GTIN,Identifikationsnummer\r\n"
        "Uhrenvergleich,,,,Station Name,19930301,00:01:00,13:12:00,,,,,,,,,,,,,,,,,crid://rabe.ch/v1/test\r\n"
        'Meme Dub,Da Composah,Da Gang,,Station Name,19930301,00:01:00,13:37:00,,DEZ650710376,,,,,,"album, but string",,,,,,,,,crid://rabe.ch/v1/test\r\n'
        'Bubbles,,"Mary\'s Surprise Act, Climmy Jiff",,Station Name,19930301,00:01:00,16:20:00,,DEZ650710376,Jane Records,,,,20221213,Da Alboom,,,,,,,,greedy-capitalist-number,crid://rabe.ch/v1/test\r\n'
        ",,Artists as string not list,,Station Name,19930301,00:01:00,17:17:17,,,,,,,,,,,,,,,,,crid://rabe.ch/v1/test\r\n"
        "Long Playing,,,,Station Name,19930301,19:48:57,18:18:18,,,,,,,,,,,,,,,,,crid://rabe.ch/v1/test\r\n"
        "composer in works,Worker,,,Station Name,19930301,19:48:57,18:18:18,,,,,,,,,,,,,,,,,crid://rabe.ch/v1/test\r\n"
        "composer better in works,composer,same,,Station Name,19930301,19:48:57,18:18:18,,,,,,,,,,,,,,,,,crid://rabe.ch/v1/test\r\n"
    )
    # pylint: enable=line-too-long
    mock_cridlib_get.assert_has_calls(
        [
            call(
                timestamp=datetime(1993, 3, 1, 13, 12, tzinfo=timezone.utc),
                fragment="acrid=a1",
            ),
            call(
                timestamp=datetime(1993, 3, 1, 13, 37, tzinfo=timezone.utc),
                fragment="acrid=a2",
            ),
            call(
                timestamp=datetime(1993, 3, 1, 16, 20, tzinfo=timezone.utc),
                fragment="acrid=a3",
            ),
            call(
                timestamp=datetime(1993, 3, 1, 17, 17, 17, tzinfo=timezone.utc),
                fragment="acrid=a4",
            ),
            call(
                timestamp=datetime(1993, 3, 1, 18, 18, 18, tzinfo=timezone.utc),
                fragment="acrid=a5",
            ),
        ]
    )


def test_get_xlsx():
    """Test get_xlsx."""

    # empty data
    data = []
    xlsx = suisa_sendemeldung.get_xlsx(data)
    workbook = load_workbook(xlsx)
    worksheet = workbook.active
    # pylint: disable=duplicate-code
    assert list(worksheet.values) == [
        (
            "Titel",
            "Komponist",
            "Interpret",
            "Interpreten-Info",
            "Sender",
            "Sendedatum",
            "Sendedauer",
            "Sendezeit",
            "Werkverzeichnisangaben",
            "ISRC",
            "Label",
            "CD ID / Katalog-Nummer",
            "Aufnahmedatum",
            "Aufnahmeland",
            "Erstveröffentlichungsdatum",
            "Titel des Tonträgers (Albumtitel)",
            "Autor Text",
            "Track Nummer",
            "Genre",
            "Programm",
            "Bestellnummer",
            "Marke",
            "Label Code",
            "EAN/GTIN",
            "Identifikationsnummer",
        )
    ]
    # pylint: enable=duplicate-code


def test_get_email_attachment():
    """Test get_email_attachment."""
    filename = "test.xlsx"
    filetype = "xlsx"
    data = BytesIO()
    part = suisa_sendemeldung.get_email_attachment(filename, filetype, data)
    assert part.get_filename() == "test.xlsx"
    assert part.get_content_type() == "application/vnd.ms-excel"

    filename = "test.csv"
    filetype = "csv"
    data = "data"
    part = suisa_sendemeldung.get_email_attachment(filename, filetype, data)
    assert part.get_filename() == "test.csv"
    assert part.get_content_type() == "text/csv"


def test_create_message():
    """Test create_message."""
    sender = "from@example.org"
    recipient = "long-arm-of-music-industry@example.com"
    subject = "subject"
    text = "text"
    filename = "/tmp/filename"
    filetype = "csv"
    data = "data"
    carbon_copy = "cc@example.org"
    bcc = "bcc@example.org"

    msg = suisa_sendemeldung.create_message(
        sender, recipient, subject, text, filename, filetype, data, carbon_copy, bcc
    )
    assert msg.get("From") == sender
    assert msg.get("To") == recipient
    assert msg.get("Subject") == subject
    assert msg.get("Cc") == carbon_copy
    assert msg.get("Bcc") == bcc


def test_send_message():
    """Test send_message."""
    msg = Message()

    # no auth
    with patch("suisa_sendemeldung.suisa_sendemeldung.SMTP", autospec=True) as mock:
        suisa_sendemeldung.send_message(msg)
        mock.assert_called_once_with("127.0.0.1")
        ctx = mock.return_value.__enter__.return_value
        ctx.starttls.assert_called_once()
        ctx.send_message.assert_called_once_with(msg)

    # auth, user provided login
    with patch("suisa_sendemeldung.suisa_sendemeldung.SMTP", autospec=True) as mock:
        suisa_sendemeldung.send_message(msg, "127.0.0.1", "user", "password")
        mock.assert_called_once_with("127.0.0.1")
        ctx = mock.return_value.__enter__.return_value
        ctx.starttls.assert_called_once()
        ctx.login.assert_called_once_with("user", "password")

    # auth, user from msg
    with patch("suisa_sendemeldung.suisa_sendemeldung.SMTP", autospec=True) as mock:
        msg.add_header("From", "test@example.org")
        suisa_sendemeldung.send_message(msg, "127.0.0.1", None, "password")
        mock.assert_called_once_with("127.0.0.1")
        ctx = mock.return_value.__enter__.return_value
        ctx.starttls.assert_called_once()
        ctx.login.assert_called_once_with("test@example.org", "password")


@mark.parametrize(
    "test_music,expected",
    [
        ({"external_ids": {"isrc": "DEZ650710376"}}, "DEZ650710376"),
        ({"external_ids": {"isrc": ["DEZ650710376"]}}, "DEZ650710376"),
        ({"external_ids": {"isrc": "DE Z65 07 10376"}}, "DEZ650710376"),
        ({"external_ids": {"isrc": "ISRCDEZ650710376"}}, "DEZ650710376"),
        ({"external_ids": {"isrc": "123456789-1"}}, ""),
        ({"isrc": "DEZ650710376"}, "DEZ650710376"),
        ({"isrc": ["DEZ650710376"]}, "DEZ650710376"),
        ({"isrc": "ISRCDEZ650710376"}, "DEZ650710376"),
        ({"isrc": "DE Z65 07 10376"}, "DEZ650710376"),
        ({"isrc": "123456789-1"}, ""),
    ],
)
def test_get_isrc(test_music, expected):
    """Test get_isrc."""

    isrc = suisa_sendemeldung.get_isrc(test_music)
    assert isrc == expected
