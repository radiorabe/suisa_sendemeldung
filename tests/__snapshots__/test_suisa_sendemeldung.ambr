# serializer version: 1
# name: test_cli_help
  '''
  usage: pytest [-h] --bearer-token BEARER_TOKEN --project-id PROJECT_ID
                --stream-id STREAM_ID [--station-name STATION_NAME]
                [--station-name-short STATION_NAME_SHORT] [--file]
                [--filetype {xlsx,csv}] [--email] [--email-from EMAIL_FROM]
                [--email-to EMAIL_TO] [--email-cc EMAIL_CC]
                [--email-bcc EMAIL_BCC] [--email-server EMAIL_SERVER]
                [--email-login EMAIL_LOGIN] [--email-pass EMAIL_PASS]
                [--email-subject EMAIL_SUBJECT] [--email-text EMAIL_TEXT]
                [--email-footer EMAIL_FOOTER]
                [--responsible-email RESPONSIBLE_EMAIL]
                [--start-date START_DATE] [--end-date END_DATE] [--last-month]
                --timezone TIMEZONE [--locale LOCALE] [--filename FILENAME]
                [--stdout]
  
  options:
    -h, --help            show this help message and exit
    --bearer-token BEARER_TOKEN
                          the bearer token for ACRCloud (required) [env var:
                          BEARER_TOKEN]
    --project-id PROJECT_ID
                          the id of the project at ACRCloud (required) [env var:
                          PROJECT_ID]
    --stream-id STREAM_ID
                          the id of the stream at ACRCloud (required) [env var:
                          STREAM_ID]
    --station-name STATION_NAME
                          Station name, used in Output and Emails [env var:
                          STATION_NAME]
    --station-name-short STATION_NAME_SHORT
                          Shortname for station as used in Filenames (locally
                          and in attachment) [env var: STATION_NAME_SHORT]
    --file                create file [env var: FILE]
    --filetype {xlsx,csv}
                          filetype to attach to email or write to file [env var:
                          FILETYPE]
    --email               send an email [env var: EMAIL]
    --email-from EMAIL_FROM
                          the sender of the email [env var: EMAIL_FROM]
    --email-to EMAIL_TO   the recipients of the email [env var: EMAIL_TO]
    --email-cc EMAIL_CC   the cc recipients of the email [env var: EMAIL_CC]
    --email-bcc EMAIL_BCC
                          the bcc recipients of the email [env var: EMAIL_BCC]
    --email-server EMAIL_SERVER
                          the smtp server to send the mail with [env var:
                          EMAIL_SERVER]
    --email-login EMAIL_LOGIN
                          the username to logon to the smtp server (default:
                          email_from) [env var: EMAIL_LOGIN]
    --email-pass EMAIL_PASS
                          the password for the smtp server [env var: EMAIL_PASS]
    --email-subject EMAIL_SUBJECT
                          Template for subject of the email. Placeholders are
                          $station_name, $year and $month. [env var:
                          EMAIL_SUBJECT]
    --email-text EMAIL_TEXT
                          Template for email text. Placeholders are
                          $station_name, $month, $year, $previous_year,
                          $responsible_email, and $email_footer. [env var:
                          EMAIL_TEXT]
    --email-footer EMAIL_FOOTER
                          Footer for the Email [env var: EMAIL_FOOTER]
    --responsible-email RESPONSIBLE_EMAIL
                          Used to hint whom to contact in the emails text. [env
                          var: RESPONSIBLE_EMAIL]
    --start-date START_DATE
                          the start date of the interval in format YYYY-MM-DD
                          (default: 30 days before end_date) [env var:
                          START_DATE]
    --end-date END_DATE   the end date of the interval in format YYYY-MM-DD
                          (default: today) [env var: END_DATE]
    --last-month          download data of whole last month [env var:
                          LAST_MONTH]
    --timezone TIMEZONE   set the timezone for localization [env var: TIMEZONE]
    --locale LOCALE       set locale for date and time formatting [env var:
                          LOCALE]
    --filename FILENAME   Output filename. Default: -
                          <station_name_short>_<year>_<month>.csv when reporting
                          last month - <station_name_short>_<start_date>.csv
                          else [env var: FILENAME]
    --stdout              also print to stdout [env var: STDOUT]
  
   In general, command-line values override environment variables which override
  defaults.
  
  '''
# ---
